src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs

```diff
@@ -454,9 +444,43 @@ internal bool IsReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAn
             return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesReferenceType(type.TypeSymbol, arg));
         }
 
-        internal bool IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
+        internal bool? IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
         {
-            return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesIsNotNullableIfReferenceType(type, arg));
+            Debug.Assert(!constraintTypes.IsDefaultOrEmpty);
+            inProgress = inProgress.Prepend(this);
+
+            bool? result = false;
+            foreach (TypeSymbolWithAnnotations constraintType in constraintTypes)
+            {
+                if (constraintType.IsAnnotated)
+                {
+                    continue;
+                }
+
+                if (constraintType.TypeKind == TypeKind.TypeParameter)
+                {
+                    bool? isNotNullableIfReferenceType = ((TypeParameterSymbol)constraintType.TypeSymbol).GetIsNotNullableIfReferenceType(inProgress);
+
+                    if (isNotNullableIfReferenceType == false)
+                    {
+                        continue;
+                    }
+                    else if (isNotNullableIfReferenceType == null)
+                    {
+                        result = null;
+                        continue;
+                    }
+                }
+
+                if (constraintType.NonNullTypesContext.NonNullTypes == true)
+                {
+                    return true;
+                }
+
+                result = null;
```

Comment:
>result = null; [](start = 16, length = 14)

Note: the test I'm asking about would cover this path (NonNullTypes(false)).

Reply:
> Note: the test I'm asking about would cover this path (NonNullTypes(false)).

I believe existing tests cover this line.

---
In reply to: [214178494](https://github.com/dotnet/roslyn/pull/29586#discussion_r214178494) [](ancestors = 214178494)